doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,331 @@
+ *  source: string | Buffer,
+ * }} ModuleExportsContext
+ */
+/**
+ * @typedef {{
+*   exports: any
+* }} ModuleExportsResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleExportsContext} context
+ * @param {(context: ModuleExportsContext) => {ModuleExportsResult}} nextExports
+ * @returns {ModuleExportsResult}
+ */
+function exports(url, context, nextExports) {
+  let exported = nextExports(url, context);
+  const stats = getStats(url);
+  if (stats.name && modules.includes(stats.name)) {
+    const newExports = userCallback(originalExports, stats.name, stats.basedir);
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+If the module loaded is CJS (`context.format` is `"commonjs"`), If the default step is run via `nextExports` run but the `exports` returned later is not reference equal to the original exports object, it will only affect later `module.exports` access in the original module, but does not affect contextual `exports.foo` accesses within the module. There are two options to address this:
+
+1. We allow the default `exports` step to take an optional `context.exports`. If it's configured, it will be used during the wrapper invocation, which unifies the contextual access. Hooks will need to use a proxy if they want the new exports object to be based on the result from default evaluation (since by the time the new overridden exports object is created, the module hasn't been evaluated yet).
+2. We expose the compiled CJS wrapper function in `context` so that users get to invoke it themselves and skip the default `exports` step.
+
+1 may provide better encapsulation, because with 2 would mean that changes to the CJS wrapper function parameters becomes breaking.
+
+If the module loaded is ESM (`context.format` is `"module"`) all the direct modification to `exports` are no-ops because ESM namespaces are not mutable. Returning a new `exports` for `require(esm)` only affects the caller of `require(esm)`, unless the hook returns a proxy to connect the changes between the new exports object and the module namespace object.
+
+## `link` (import-only): from source code to compiled module records
+
+This needs to work with experimental vm modules for passing module instances around.
+
+Example mock of import-in-the-middle:
+
+```js
+/**
+ * @typedef {{
+ *  specifier: string,
+ *  source: string | Buffer,
+ * }} ModuleLinkContext
+ */
+/**
+ * @typedef {{
+*   module: vm.Module
+* }} ModuleLinkResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleLinkContext} context
+ * @param {(context: ModuleLinkContext) => {ModuleLinkResult}} nextLink
+ * @returns {ModuleLinkResult}
+ */
+function link(url, context, nextLink) {
+  const linked = nextLink(url, context);
+  if (!shouldOverride(url, context.specifier)) {
+    return linked;
+  }
+  assert.strictEqual(linked.module.status, 'linked');  // Original module is linked at this point
+  let source = `import * as original from 'original';`;
```

Comment:
Actually I noticed that with this and a cache manipulation API and `require('module')`, this can support hot module replacement - it needs `require('module')` because the original module must not be linked into the graph, but dynamically loaded, for the originals to be swapped out. That would be limited to modules without TLA but I imagine there are already other implications with hot reload of a module that does a TLA that make it dangerous...

Reply:
`require('module')` does link the required module into the module graph, same as `import('module')`. HMR requires a mechanism to unload the module from the module graph.

Reply:
It doesn’t actually link the required modules with the module that does the require - they are only independent graphs with a one way reference, which is only established at evaluation time of the parent module, not at link time of the parent module, so we can safely offload the child module and modify the reference that is only created at evaluation time; the link-time reference, however, is immutable, and that’s what I meant - that the facade module has to require the original module in the graph, not importing it statically.

Reply:
I see. I think this also applies to `import('module')` calls? The original paragraph reads to me like it only applies to `require('module')`.

Reply:
import would be async so it’s going to be racy